### PR TITLE
Associates ldap_dn on a first User() login

### LINKED ISSUE #507: LDAP login works only once per user
##### ISSUE TYPE
 - Bug Report

##### SUMMARY
After configuring LDAP, I'm able to login. Attributes and groups seem to sync correctly, After logging out however I can no longer login with that same username. The only error in the `awx-task` log is `awx.api.generics Login failed for user`.

##### ENVIRONMENT
* AWX version: 19.3.0
* Operator version: 0.12.0
* Kubernetes version: 1.20.0-1089
* AWX install method: OKD 4.7

##### STEPS TO REPRODUCE

* Install AWX and configure LDAP authentication as below.
* Login with an LDAP user
* Logout
* Login again

##### EXPECTED RESULTS

User should be able to login again successfully.

##### ACTUAL RESULTS

User can only login once.

##### ADDITIONAL INFORMATION

Redacted config as follows:
```
LDAP Server URI
ldap://ldap-server.example.com/
LDAP Bind DN
uid=svc_awx,cn=sysaccounts,cn=etc,dc=example,dc=com
LDAP Bind Password
Encrypted
LDAP User DN Template
uid=%(user)s,cn=users,cn=accounts,dc=example,dc=com
LDAP Group Type
GroupOfNamesType
LDAP Require Group
cn=awx-users,cn=groups,cn=accounts,dc=example,dc=com
LDAP Deny Group
Not configured
LDAP Start TLS
On

LDAP User Search
[
  "cn=users,cn=accounts,dc=example,dc=com",
  "SCOPE_SUBTREE",
  "(uid=%(user)s)"
]

LDAP Group Search
[
  "cn=groups,cn=accounts,dc=example,dc=com",
  "SCOPE_SUBTREE",
  "(objectClass=nestedGroup)"
]

LDAP User Attribute Map
{
  "first_name": "givenName",
  "last_name": "sn",
  "email": "mail"
}

LDAP User Flags By Group
{
  "is_superuser": [
    "cn=awx-admins,cn=groups,cn=accounts,dc=example,dc=com"
  ]
}
```

Deleting the LDAP user using the admin account seems to allow logging in again.

### PR DESCRIPTION
**Pull request by tchellomello**
##### SUMMARY
Related: https://github.com/ansible/awx/issues/10883
Fixes: https://github.com/ansible/awx-operator/issues/507

Related: https://github.com/ansible/awx/pull/9703

Fix the missing ldap_dn attribute for new LDAP users.

To avoid calling the user.save() on every single login (PR#9703)
we can check if the user.profile is available. For new users,
accessing the user.profile throws an ValueError exception which
is capture on this fix.

* Example:
```
>>> _ = user.profile
*** ValueError: save() prohibited to prevent data loss due to unsaved related object 'user'.
>>> User.objects.filter(username=user.username).count()
0
```

This way, the user.save() gets called for brand users and will get the
ldap_dn associated as expected.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME

 - API

##### AWX VERSION

```
19.3.0
```

##### ADDITIONAL INFORMATION
image

### CODE REVIEW

**Comment by rooftopcellist** [comment]
I checked this PR out and I can confirm that I can:

1. Configure LDAP server (ldap.forumsys.com)
2. Log in with an LDAP user

image

3. Make them a sys admin
image

4. Log back in as the LDAP user

image

**Review by rooftopcellist — APPROVED** [review]
LGTM

File: awx/sso/backends.py [hunk 1]
```diff
@@ -395,10 +395,17 @@ def on_populate_user(sender, **kwargs):
         remove = bool(team_opts.get('remove', True))
         _update_m2m_from_groups(user, ldap_user, team.member_role.members, users_opts, remove)
 
+    # Check if user.profile is available, otherwise force user.save()
+    try:
+        _ = user.profile
+    except ValueError:
+        force_user_update = True
+    finally:
+        if force_user_update:
```
Comment (chrismeyersfsu) [inline awx/sso/backends.py:404]:
```
>>> try:
...   raise Exception
... except ValueError:
...   print("hello")
... finally:
...   print("good bye")
...
good bye
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
Exception
```

You can end up in the `finally` due to some exception other than `ValueError` & `force_user_update` will be an undefined variable. Set `force_user_update = False` before the `try:` to be defensive so that if a programmer comes here in the future they are less likely to add code to the `try:` that might increase the chance of raising some other uncaught exception.

TL;DR set `force_user_update = False` before the `try:`.

File: awx/sso/backends.py [hunk 2]
```diff
@@ -395,10 +395,17 @@ def on_populate_user(sender, **kwargs):
...
+            user.save()
+
     # Update user profile to store LDAP DN.
-    if force_user_update:
-        user.save()
-        profile = user.profile
-        if profile.ldap_dn != ldap_user.dn:
-            profile.ldap_dn = ldap_user.dn
-            profile.save()
+    profile = user.profile
```
Comment (chrismeyersfsu) [inline awx/sso/backends.py:408]: Can this `user.profile` result in a `ValueError`? Or will the `save()` above ensure that `user.profile` gets set?

**Review by chrismeyersfsu — APPROVED** [review]

File: awx/sso/backends.py (hunk 1 above)
Comment (tchellomello) [reply awx/sso/backends.py:404]: @chrismeyersfsu the `force_user_update = False` is already set before the `try:`  statement at https://github.com/ansible/awx/blob/devel/awx/sso/backends.py#L357

File: awx/sso/backends.py (hunk 2 above)
Comment (tchellomello) [reply awx/sso/backends.py:408]: @chrismeyersfsu  during my tests, the only time I was able to get a `ValueError` was for a brand new user. For users already present in the database,  `user.profile`  worked as expected.

**Review by chrismeyersfsu — APPROVED** [review]

**Comment by damluji** [comment]
Any updates when a release with this fix is coming out please?
